strum_macros/src/macros/enum_table.rs

```diff
@@ -0,0 +1,193 @@
+        }
+
+        // Error on variants with data
+        let Fields::Unit = &variant.fields else {
+            return Err(syn::Error::new(
+                variant.fields.span(),
+                "`EnumTable` doesn't support enums with non-unit variants",
+            ))
+        };
+
+        let pascal_case = &variant.ident;
+        let snake_case = format_ident!("_{}", snakify(&pascal_case.to_string()));
+
+        get_matches.push(quote! {#name::#pascal_case => &self.#snake_case,});
+        get_matches_mut.push(quote! {#name::#pascal_case => &mut self.#snake_case,});
+        set_matches.push(quote! {#name::#pascal_case => self.#snake_case = new_value,});
+        closure_fields.push(quote! {#snake_case: func(#name::#pascal_case),});
+        transform_fields.push(quote! {#snake_case: func(#name::#pascal_case, &self.#snake_case),});
+        pascal_idents.push(pascal_case);
+        snake_idents.push(snake_case);
+    }
+
+    // Error on empty enums
+    if pascal_idents.is_empty() {
+        return Err(syn::Error::new(
+            variants.span(),
+            "`EnumTable` requires at least one non-disabled variant",
+        ));
+    }
+
+    // if the index operation can panic, add that to the documentation
+    if !disabled_variants.is_empty() {
+        doc_comment.push_str(&format!(
+            "\n# Panics\nIndexing `{table_name}` with any of the following variants will cause a panic:"
+        ));
+        for variant in disabled_variants {
+            doc_comment.push_str(&format!("\n\n- `{name}::{variant}`"));
+        }
+    }
+
+    let doc_new = format!("Create a new {table_name} with a value for each variant of {name}");
+    let doc_closure =
+        format!("Create a new {table_name} by running a function on each variant of `{name}`");
+    let doc_transform = format!("Create a new `{table_name}` by running a function on each variant of `{name}` and the corresponding value in the current `{table_name}`");
+    let doc_filled = format!("Create a new `{table_name}` with the same value in each field.");
+    let doc_option_all = format!("Converts `{table_name}<Option<T>>` into `Option<{table_name}<T>>`. Returns `Some` if all fields are `Some`, otherwise returns `None`.");
+    let doc_result_all_ok = format!("Converts `{table_name}<Result<T, E>>` into `Result<{table_name}, E>`. Returns `Ok` if all fields are `Ok`, otherwise returns `Err`.");
+
+    Ok(quote! {
+        #[doc = #doc_comment]
+        #[allow(
+            missing_copy_implementations,
+        )]
+        #[derive(Debug, Clone, Default, PartialEq, Eq, Hash)]
+        #vis struct #table_name<T> {
+            #(#snake_idents: T,)*
+        }
+
+        impl<T: Clone> #table_name<T> {
+            #[doc = #doc_filled]
+            #vis fn filled(value: T) -> #table_name<T> {
+                #table_name {
+                    #(#snake_idents: value.clone(),)*
+                }
+            }
+        }
+
+        impl<T> #table_name<T> {
```

Comment:
Can we add a `get` and `get_mut` methods to compliment Index and match the semantics of arrays, Vecs and HashMaps?

Reply:
I don't believe we need these. The main reason to have both `Index` and `get` for existing types is that a key doesn't always map to a value. For `EnumTable`, those methods wouldn't be any different from just using `Index` and `IndexMut`. Again, I don't care much on this, just want to make sure we're on the same page.